### PR TITLE
2FA: Version 2 (Too Too Much Authentication)

### PR DESCRIPTION
**Pull request by sampoder**
Oh too much. How this works is described in https://github.com/hackclub/hcb/pull/6356.

### CODE REVIEW

File: app/controllers/logins_controller.rb [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  def new
+    @prefill_email = params[:email] if params[:email].present?
+    @return_to = params[:return_to]
+  end
+
+  # when you submit your email
+  def create
+    email = params[:email]
+    user = User.find_by(email:)
+    login = user.logins.create
```
Comment (malted) [inline app/controllers/logins_controller.rb]: When user is `nil`, `.logins` throws

Comment (sampoder) [reply app/controllers/logins_controller.rb]: fixed! thanks for flagging this

File: app/controllers/logins_controller.rb [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    @return_to = params[:return_to]
+  end
+
+  # when you submit your email
+  def create
+    email = params[:email]
+    user = User.find_or_create_by!(email:)
+    login = user.logins.create
+
+    has_webauthn_enabled = user&.webauthn_credentials&.any?
```
Comment (garyhtou) [inline app/controllers/logins_controller.rb]: see if you can make this code here generic so that we don't have conditions for specific login methods. what might help with this is a list of the user's configured login methods

File: app/controllers/logins_controller.rb [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    else
+      session[:auth_email] = email
+      redirect_to choose_login_preference_login_path(login, return_to: params[:return_to])
+    end
+  end
+
+  def choose_login_preference
+    @email = session[:auth_email]
+    @user = User.find_by_email(@email)
+    @login = params[:id] ? Login.find_by_hashid!(params[:id]) : @user.logins.create
```
Comment (garyhtou) [inline app/controllers/logins_controller.rb]: create a `set_login`.
```ruby
@login = Login.find(params[:id])
same_browser = secure_compare(@login.browser_token, cookies[:browser_token]
unless same_browser
  # show error that it's a different browser
end
```

File: app/controllers/logins_controller.rb [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+    @return_to = params[:return_to]
+    return redirect_to auth_users_path if @email.nil?
+
+    session.delete :login_preference
+  end
+
+  def set_login_preference
+    @login = Login.find_by_hashid!(params[:id])
+    @email = params[:email]
```
Comment (garyhtou) [inline app/controllers/logins_controller.rb]: Use `login.user.email`

File: app/controllers/logins_controller.rb [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+      session[:login_preference] = "email" if remember
+      redirect_to login_code_login_path(@login), status: :temporary_redirect
+    when "webauthn"
+      # This should never happen, because WebAuthn auth is handled on the frontend
+      redirect_to choose_login_preference_login_path(@login)
+    end
+  end
+
+  # post to request login code
+  def login_code
```
Comment (garyhtou) [inline app/controllers/logins_controller.rb]: see if you can make a generic submit endpoint

File: app/models/login.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+#
+# Indexes
+#
+#  index_logins_on_session_id  (session_id)
+#  index_logins_on_user_id     (user_id)
+#
+class Login < ApplicationRecord
+  include AASM
+  include Hashid::Rails
+  belongs_to :user, optional: true
```
Comment (garyhtou) [inline app/models/login.rb]: this is no longer needed. it should be required
```suggestion
  belongs_to :user
```

File: app/models/login.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  belongs_to :session, optional: true
```
Comment (garyhtou) [inline app/models/login.rb]:
```suggestion
  belongs_to :session, optional: true
  validate do
    if sessions.present? && !completed?
      # how did we create session when it's not complete?!
    end
  end
```

Comment (garyhtou) [inline app/models/login.rb]:
```suggestion
  belongs_to :session, optional: true
  validate do
    if session.user != user
      # huh?!!
    end
  end
```

File: app/models/login.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  has_paper_trail
+  store :authentication_factors, accessors: [:sms, :email, :webauthn], prefix: :authenticated_with
+  aasm do
+    state :pending, initial: true
```
Comment (garyhtou) [inline app/models/login.rb]:
```suggestion
    state :incomplete, initial: true
```

File: app/views/logins/login_code.html.erb [hunk 9]
```diff
@@ -11,11 +11,18 @@
 
   <%= render partial: "users/sms_auth_notice" %>
 
-  <p>We just sent a login code to <%= @use_sms_auth ? "your phone number ending in: " + @phone_last_four : @email %></p>
+  <p>
+    <% if @login.authentication_factors.values.count(true) > 0 %>
```
Comment (garyhtou) [inline app/views/logins/login_code.html.erb]: I see that we're using `@login.authentication_factors.values.count(true)` multiple times. put that in a method

File: db/migrate/20240611042358_create_login_attempts_table.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class CreateLoginAttemptsTable < ActiveRecord::Migration[7.1]
+  def change
+    create_table :logins do |t|
+      t.belongs_to :user
+      t.belongs_to :session, null: true
```
Comment (garyhtou) [inline db/migrate/20240611042358_create_login_attempts_table.rb]:
```suggestion
      t.belongs_to :user, null: false
      t.belongs_to :session
```

File: db/schema.rb [hunk 11]
```diff
@@ -10,8 +10,9 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema[7.1].define(version: 2024_05_31_075659) do
+ActiveRecord::Schema[7.1].define(version: 2024_06_11_042358) do
   # These are extensions that must be enabled in order to support this database
+  enable_extension "amcheck"
```
Comment (garyhtou) [inline db/schema.rb]:
```suggestion
```

File: db/schema.rb [hunk 12]
```diff
@@ -2048,6 +2061,7 @@
     t.string "payout_method_type"
     t.bigint "payout_method_id"
     t.integer "comment_notifications", default: 0, null: false
+    t.integer "charge_notifications", default: 0, null: false
```
Comment (sampoder) [inline db/schema.rb]:
```suggestion
```

File: db/schema.rb [hunk 13]
```diff
@@ -588,10 +589,10 @@
     t.datetime "created_at", precision: nil, null: false
     t.datetime "updated_at", precision: nil, null: false
     t.text "slug"
+    t.datetime "deleted_at", precision: nil
     t.datetime "archived_at"
     t.bigint "archived_by_id"
     t.string "aasm_state"
```
Comment (sampoder) [inline db/schema.rb]:
```suggestion
    t.text "slug"
    t.datetime "archived_at"
    t.bigint "archived_by_id"
    t.string "aasm_state"
    t.datetime "deleted_at", precision: nil
```

File: db/schema.rb [hunk 14]
```diff
@@ -10,8 +10,9 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema[7.1].define(version: 2024_06_13_205428) do
+ActiveRecord::Schema[7.1].define(version: 2024_06_17_180108) do
   # These are extensions that must be enabled in order to support this database
+  enable_extension "amcheck"
```
Comment (sampoder) [inline db/schema.rb]:
```suggestion
```

File: db/schema.rb [hunk 15]
```diff
@@ -52,9 +53,9 @@
     t.bigint "payment_recipient_id"
     t.string "recipient_email"
     t.boolean "send_email_notification", default: false
+    t.boolean "same_day", default: false, null: false
     t.string "company_name"
     t.string "company_entry_description"
```
Comment (sampoder) [inline db/schema.rb]:
```suggestion
    t.string "company_name"
    t.string "company_entry_description"
    t.boolean "same_day", default: false, null: false
```

File: app/models/login.rb [hunk 16]
```diff
@@ -9,35 +9,60 @@
...
+  validate do
+    if user_session.present? && !complete?
+      # how did we create session when it's not complete?!
+      errors.add(:base, "An incomplete login has a session present.")
+    end
+  end
+
+  validate do
+    if user_session.present? && user_session.user != user
+      # how did we create session when it's not complete?!
```
Comment (garyhtou) [inline app/models/login.rb]:
```suggestion
```

File: app/models/login.rb [hunk 17]
```diff
@@ -9,35 +9,60 @@
...
+      errors.add(:base, "A login with a session present has a session.user / user mismatch.")
+    end
+  end
```
Comment (garyhtou) [inline app/models/login.rb]: Let's airbrake these

File: app/controllers/logins_controller.rb [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+    redirect_to auth_users_path, flash: { error: "Something went wrong." }
+  rescue ActiveRecord::RecordInvalid => e
+    redirect_to auth_users_path, flash: { error: e.record.errors&.full_messages&.join(". ") }
+  end
+
+  private
+
+  def set_login
+    if params[:id]
+      @login = Login.find_by_hashid!(params[:id])
```
Comment (garyhtou) [inline app/controllers/logins_controller.rb]:
```suggestion
      @login = Login.incomplete.find_by_hashid!(params[:id])
```

File: app/controllers/logins_controller.rb [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+        @login.update(authenticated_with_totp: true)
+      else
+        return redirect_to totp_users_path(email: @user.email), flash: { error: "Invalid TOTP code, please try again." }
+      end
+    end
+
+    # Clear the flash - this prevents the error message showing up after an unsuccessful -> successful login
+    flash.clear
+
+    if @login.complete?
```
Comment (garyhtou) [inline app/controllers/logins_controller.rb]:
```suggestion
    # Only create a user session if authentication factors are met AND this login
    # has not created a user session before
    if @login.complete? && @login.user_session.nil?
```

File: app/controllers/logins_controller.rb [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+    if params[:id]
+      @login = Login.find_by_hashid!(params[:id])
+      unless valid_browser_token?
+        # error! browser token doesn't match the cookie.
+        flash[:error] = "This doesn't seem to be the browser who began this login; please ensure cookies are enabled."
+        redirect_to auth_users_path
+      end
+    elsif session[:auth_email]
+      @login = User.find_by_email(session[:auth_email]).logins.create
+      cookies.signed[:browser_token] = @login.browser_token
```
Comment (garyhtou) [inline app/controllers/logins_controller.rb]:
```suggestion
      cookies.signed[:browser_token] = @login.browser_token
    else
      raise ActionController::ParameterMissing.new("Missing login")
```

File: app/controllers/users_controller.rb
Comment (garyhtou) [inline app/controllers/users_controller.rb:1]: seems to have unrelated changes

File: app/javascript/controllers/webauthn_auth_controller.js [hunk 21]
```diff
@@ -59,14 +60,20 @@ export default class extends Controller {
...
-          this.hasRememberInputTarget && this.rememberInputTarget.checked,
-        ...(await this.fingerprint()),
-      })
+      submitForm(
+        this.loginIdValue
+          ? `/logins/${this.loginIdValue}/complete`
+          : `/logins/complete`,
+        {
+          credential: JSON.stringify(credential),
+          email: loginEmail,
```
Comment (garyhtou) [inline app/javascript/controllers/webauthn_auth_controller.js]:
```suggestion
```

File: app/models/login.rb [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+#
+# Table name: logins
+#
+#  id                     :bigint           not null, primary key
+#  aasm_state             :string
+#  authentication_factors :jsonb
+#  created_at             :datetime         not null
+#  updated_at             :datetime         not null
+#  user_id                :bigint           not null
+#  user_session_id        :bigint
```
Comment (garyhtou) [inline app/models/login.rb]: missing annotation for browser token

**Review by garyhtou — COMMENTED** [review]
- Can you write a test that ensure creating a new Login results in an incomplete state.

File: app/controllers/logins_controller.rb [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+    redirect_to auth_users_path, flash: { error: "Something went wrong." }
+  rescue ActiveRecord::RecordInvalid => e
+    redirect_to auth_users_path, flash: { error: e.record.errors&.full_messages&.join(". ") }
+  end
+
+  private
+
+  def set_login
+    if params[:id]
+      @login = Login.find_by_hashid!(params[:id])
```
Comment (sampoder) [inline app/controllers/logins_controller.rb]:
```suggestion
      @login = Login.incomplete.find_by_hashid!(params[:id])
```

File: app/controllers/logins_controller.rb [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+
+  def initialize_sms_params
+    return if @force_use_email && !@login.authenticated_with_email
+    return if @login.authenticated_with_sms
+
+    if @login.user&.use_sms_auth || @login.authenticated_with_email
+      @use_sms_auth = true
+      @phone_last_four = @login.user.phone_number.last(4)
+    end
+  end
```
Comment (sampoder) [inline app/controllers/logins_controller.rb]: Add a check that the use has a verified phone number

File: app/controllers/logins_controller.rb [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+    redirect_to auth_users_path, flash: { error: "Something went wrong." }
+  rescue ActiveRecord::RecordInvalid => e
+    redirect_to auth_users_path, flash: { error: e.record.errors&.full_messages&.join(". ") }
+  end
+
+  private
+
+  def set_login
+    if params[:id]
+      @login = Login.incomplete.find_by_hashid!(params[:id])
```
Comment (sampoder) [inline app/controllers/logins_controller.rb]: redirect to auth page

Comment (sampoder) [reply app/controllers/logins_controller.rb]: rescue record not found

File: app/controllers/logins_controller.rb [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+      timezone: params[:timezone],
+      ip: request.remote_ip
+    }
+  end
+
+  def initialize_sms_params
+    return if @force_use_email && !@login.authenticated_with_email
+    return if @login.authenticated_with_sms
+
+    if @login.user&.use_sms_auth || @login.authenticated_with_email
```
Comment (sampoder) [inline app/controllers/logins_controller.rb]:
```suggestion
    if @login.user&.use_sms_auth || (@login.authenticated_with_email && @login.user&.phone_number_verified)
```

File: app/models/login.rb [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+#
+# Indexes
+#
+#  index_logins_on_user_id          (user_id)
+#  index_logins_on_user_session_id  (user_session_id)
+#
+class Login < ApplicationRecord
```
Comment (garyhtou) [inline app/models/login.rb]: Make Login expire at 15 mins
- Add a constant with 15.minutes
- Add a guard to transition into `complete`
- Add a scope called `active`
- Use the `active` when finding for Login
- In the controller, when we set the browser token, use the constant to expire it

File: app/controllers/users_controller.rb [hunk 28]
```diff
@@ -603,17 +387,7 @@ def user_params
...
-    user = User.find_by(email: @email)
-    if user&.use_sms_auth
-      @use_sms_auth = true
-      @phone_last_four = user.phone_number.last(4)
-    end
-  end
-
-  # HCB used to run on bank.hackclub.com— this ensures that any old references to `bank.` URLs are translated into `hcb.`
+  # HCB used to run on bank.hackclub.com— this ensures that any old references to `bank.` URLs are translated into `hcb.`
   def migrate_return_to
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]: we can remove this method now. i think it's unused

File: app/models/login.rb [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+
+  has_paper_trail
+
+  validate do
+    if user_session.present? && !complete?
+      # how did we create session when it's not complete?!
+      Airbrake.notify("An incomplete login #{id} has a session #{session.id} present.")
+      errors.add(:base, "An incomplete login has a session present.")
+    end
+  end
```
Comment (garyhtou) [inline app/models/login.rb]: Also validate you can't have a complete without a user_session

**Review by garyhtou — COMMENTED** [review]
🪄
